File: sdk/src/riverpool.ts
/**
 * RiverPool SDK Client
 * TypeScript client for interacting with RiverPool liquidity pools
 */

import axios, { AxiosInstance } from 'axios';

// Types
export interface Pool {
  poolId: string;
  poolType: 'foundation' | 'main' | 'community';
  name: string;
  description: string;
  status: 'active' | 'paused' | 'closed';
  totalDeposits: string;
  totalShares: string;
  nav: string;
  highWaterMark: string;
  currentDrawdown: string;
  ddGuardLevel: 'normal' | 'warning' | 'reduce' | 'halt';
  minDeposit: string;
  maxDeposit: string;
  lockPeriodDays: number;
  redemptionDelayDays: number;
  dailyRedemptionLimit: string;
  seatsAvailable?: number;
  createdAt: number;
  updatedAt: number;
}

export interface Deposit {
  depositId: string;
  poolId: string;
  depositor: string;
  amount: string;
  shares: string;
  navAtDeposit: string;
  depositedAt: number;
  unlockAt: number;
  pointsEarned: string;
  isLocked: boolean;
}

export interface Withdrawal {
  withdrawalId: string;
  poolId: string;
  withdrawer: string;
  sharesRequested: string;
  sharesRedeemed: string;
  amountReceived: string;
  navAtRequest: string;
  status: 'pending' | 'processing' | 'completed' | 'cancelled';
  requestedAt: number;
  availableAt: number;
  completedAt: number;
  isReady: boolean;
}

export interface PoolStats {
  poolId: string;
  totalValueLocked: string;
  totalDepositors: number;
  totalPendingWithdrawals: string;
  realizedPnl: string;
  unrealizedPnl: string;
  totalFeesCollected: string;
  return1d: string;
  return7d: string;
  return30d: string;
  returnAllTime: string;
  updatedAt: number;
}

export interface DDGuardState {
  poolId: string;
  level: string;
  peakNav: string;
  currentNav: string;
  drawdownPercent: string;
  maxExposureLimit: string;
  triggeredAt: number;
  lastCheckedAt: number;
}

export interface NAVHistory {
  poolId: string;
  nav: string;
  totalValue: string;
  timestamp: number;
}

export interface UserPoolBalance {
  shares: string;
  value: string;
  costBasis: string;
  unrealizedPnl: string;
  pnlPercent: string;
  unlockAt: number;
  canWithdraw: boolean;
}

export interface DepositEstimate {
  shares: string;
  nav: string;
  sharePrice: string;
}

export interface WithdrawalEstimate {
  amount: string;
  nav: string;
  availableAt: number;
  queuePosition: string;
  mayBeProrated: boolean;
}

export interface DepositResponse {
  depositId: string;
  sharesReceived: string;
  navAtDeposit: string;
  unlockAt: number;
}

export interface WithdrawalRequestResponse {
  withdrawalId: string;
  sharesRequested: string;
  estimatedAmount: string;
  availableAt: number;
  queuePosition: string;
}

export interface WithdrawalClaimResponse {
  amountReceived: string;
  sharesRedeemed: string;
  remainingShares: string;
}

export interface WithdrawalCancelResponse {
  sharesReturned: string;
}

export interface RevenueStats {
  poolId: string;
  totalRevenue: string;
  spreadRevenue: string;
  fundingRevenue: string;
  liquidationProfit: string;
  tradingPnl: string;
  feeRebates: string;
  return1d: string;
  return7d: string;
  return30d: string;
  lastUpdated: number;
}

export interface RevenueRecord {
  recordId: string;
  poolId: string;
  source: 'spread' | 'funding' | 'liquidation' | 'trading' | 'fees';
  amount: string;
  navImpact: string;
  timestamp: number;
  blockHeight: number;
  marketId?: string;
  details?: string;
}

export interface RevenueBreakdown {
  poolId: string;
  period: string;
  totalAmount: string;
  breakdown: {
    spread: string;
    funding: string;
    liquidation: string;
    trading: string;
    fees: string;
  };
}

export interface RiverpoolClientConfig {
  baseUrl: string;
  timeout?: number;
}

/**
 * RiverPool Client
 * Provides methods for interacting with RiverPool liquidity pools
 */
export class RiverpoolClient {
  private client: AxiosInstance;

  constructor(config: RiverpoolClientConfig) {
    this.client = axios.create({
      baseURL: config.baseUrl,
      timeout: config.timeout || 30000,
      headers: {
        'Content-Type': 'application/json',
      },
    });
  }

  // ============ Pool Queries ============

  /**
   * Get all pools
   */
  async getPools(offset = 0, limit = 20): Promise<{ pools: Pool[]; total: number }> {
    const response = await this.client.get('/v1/riverpool/pools', {
      params: { offset, limit },
    });
    return response.data;
  }

  /**
   * Get a single pool by ID
   */
  async getPool(poolId: string): Promise<Pool> {
    const response = await this.client.get(`/v1/riverpool/pools/${poolId}`);
    return response.data;
  }

  /**
   * Get pools by type
   */
  async getPoolsByType(
    poolType: 'foundation' | 'main' | 'community'
  ): Promise<{ pools: Pool[] }> {
    const response = await this.client.get(`/v1/riverpool/pools/type/${poolType}`);
    return response.data;
  }

  /**
   * Get pool statistics
   */
  async getPoolStats(poolId: string): Promise<PoolStats> {
    const response = await this.client.get(`/v1/riverpool/pools/${poolId}/stats`);
    return response.data;
  }

  /**
   * Get DDGuard state for a pool
   */
  async getDDGuardState(poolId: string): Promise<DDGuardState> {
    const response = await this.client.get(`/v1/riverpool/pools/${poolId}/ddguard`);
    return response.data;
  }

  /**
   * Get NAV history for a pool
   */
  async getNAVHistory(
    poolId: string,
    from?: number,
    to?: number
  ): Promise<{ history: NAVHistory[] }> {
    const params: Record<string, string> = {};
    if (from) params.from = from.toString();
    if (to) params.to = to.toString();

    const response = await this.client.get(`/v1/riverpool/pools/${poolId}/nav/history`, {
      params,
    });
    return response.data;
  }

  // ============ User Queries ============

  /**
   * Get user deposits
   */
  async getUserDeposits(
    user: string
  ): Promise<{ deposits: Deposit[]; totalValue: string }> {
    const response = await this.client.get(`/v1/riverpool/user/${user}/deposits`);
    return response.data;
  }

  /**
   * Get user withdrawals
   */
  async getUserWithdrawals(user: string): Promise<{ withdrawals: Withdrawal[] }> {
    const response = await this.client.get(`/v1/riverpool/user/${user}/withdrawals`);
    return response.data;
  }

  /**
   * Get user's balance in a pool
   */
  async getUserPoolBalance(poolId: string, user: string): Promise<UserPoolBalance> {
    const response = await this.client.get(
      `/v1/riverpool/pools/${poolId}/user/${user}/balance`
    );
    return response.data;
  }

  // ============ Pool Queries ============

  /**
   * Get deposits in a pool
   */
  async getPoolDeposits(
    poolId: string,
    offset = 0,
    limit = 20
  ): Promise<{ deposits: Deposit[]; total: number }> {
    const response = await this.client.get(`/v1/riverpool/pools/${poolId}/deposits`, {
      params: { offset, limit },
    });
    return response.data;
  }

  /**
   * Get pending withdrawals for a pool
   */
  async getPendingWithdrawals(poolId: string): Promise<{
    withdrawals: Withdrawal[];
    totalPendingShares: string;
    totalPendingValue: string;
    dailyLimitRemaining: string;
  }> {
    const response = await this.client.get(
      `/v1/riverpool/pools/${poolId}/withdrawals/pending`
    );
    return response.data;
  }

  // ============ Estimation ============

  /**
   * Estimate shares for a deposit
   */
  async estimateDeposit(poolId: string, amount: string): Promise<DepositEstimate> {
    const response = await this.client.get(
      `/v1/riverpool/pools/${poolId}/estimate/deposit`,
      {
        params: { amount },
      }
    );
    return response.data;
  }

  /**
   * Estimate amount for a withdrawal
   */
  async estimateWithdrawal(poolId: string, shares: string): Promise<WithdrawalEstimate> {
    const response = await this.client.get(
      `/v1/riverpool/pools/${poolId}/estimate/withdrawal`,
      {
        params: { shares },
      }
    );
    return response.data;
  }

  // ============ Transactions ============

  /**
   * Deposit into a pool
   */
  async deposit(
    depositor: string,
    poolId: string,
    amount: string,
    inviteCode?: string
  ): Promise<DepositResponse> {
    const response = await this.client.post('/v1/riverpool/deposit', {
      depositor,
      pool_id: poolId,
      amount,
      invite_code: inviteCode,
    });
    return response.data;
  }

  /**
   * Request a withdrawal
   */
  async requestWithdrawal(
    withdrawer: string,
    poolId: string,
    shares: string
  ): Promise<WithdrawalRequestResponse> {
    const response = await this.client.post('/v1/riverpool/withdrawal/request', {
      withdrawer,
      pool_id: poolId,
      shares,
    });
    return response.data;
  }

  /**
   * Claim a withdrawal
   */
  async claimWithdrawal(
    withdrawer: string,
    withdrawalId: string
  ): Promise<WithdrawalClaimResponse> {
    const response = await this.client.post('/v1/riverpool/withdrawal/claim', {
      withdrawer,
      withdrawal_id: withdrawalId,
    });
    return response.data;
  }

  /**
   * Cancel a withdrawal
   */
  async cancelWithdrawal(
    withdrawer: string,
    withdrawalId: string
  ): Promise<WithdrawalCancelResponse> {
    const response = await this.client.post('/v1/riverpool/withdrawal/cancel', {
      withdrawer,
      withdrawal_id: withdrawalId,
    });
    return response.data;
  }

  // ============ Revenue Queries ============

  /**
   * Get pool revenue statistics
   */
  async getPoolRevenue(poolId: string): Promise<RevenueStats> {
    const response = await this.client.get(`/v1/riverpool/pools/${poolId}/revenue`);
    return response.data;
  }

  /**
   * Get revenue records for a pool
   */
  async getRevenueRecords(
    poolId: string,
    from?: number,
    to?: number,
    limit?: number
  ): Promise<{ poolId: string; records: RevenueRecord[]; total: number }> {
    const params: Record<string, string> = {};
    if (from) params.from = from.toString();
    if (to) params.to = to.toString();
    if (limit) params.limit = limit.toString();

    const response = await this.client.get(`/v1/riverpool/pools/${poolId}/revenue/records`, {
      params,
    });
    return response.data;
  }

  /**
   * Get revenue breakdown by source
   */
  async getRevenueBreakdown(
    poolId: string,
    period: '1d' | '7d' | '30d' | 'all' = '7d'
  ): Promise<RevenueBreakdown> {
    const response = await this.client.get(`/v1/riverpool/pools/${poolId}/revenue/breakdown`, {
      params: { period },
    });
    return response.data;
  }

  // ============ Utility Methods ============

  /**
   * Get Foundation LP pool
   */
  async getFoundationPool(): Promise<Pool | null> {
    try {
      return await this.getPool('foundation-lp');
    } catch {
      return null;
    }
  }

  /**
   * Get Main LP pool
   */
  async getMainPool(): Promise<Pool | null> {
    try {
      return await this.getPool('main-lp');
    } catch {
      return null;
    }
  }

  /**
   * Check if Foundation LP has available seats
   */
  async hasFoundationSeats(): Promise<boolean> {
    const pool = await this.getFoundationPool();
    return pool !== null && pool.seatsAvailable !== undefined && pool.seatsAvailable > 0;
  }

  /**
   * Calculate total user value across all pools
   */
  async getUserTotalValue(user: string): Promise<string> {
    const { deposits, totalValue } = await this.getUserDeposits(user);
    return totalValue;
  }
}

/**
 * Create a RiverPool client instance
 */
export function createRiverpoolClient(baseUrl: string, timeout?: number): RiverpoolClient {
  return new RiverpoolClient({ baseUrl, timeout });
}

export default RiverpoolClient;

File: api/types/riverpool.go
package types

import (
	"cosmossdk.io/math"
)

// RiverpoolService defines the interface for RiverPool operations
type RiverpoolService interface {
	// Pool queries
	GetPools() ([]*PoolInfo, error)
	GetPool(poolID string) (*PoolInfo, error)
	GetPoolsByType(poolType string) ([]*PoolInfo, error)
	GetPoolStats(poolID string) (*PoolStats, error)
	GetNAVHistory(poolID string, days int) ([]*NAVPoint, error)
	GetDDGuardState(poolID string) (*DDGuardState, error)

	// User queries
	GetUserDeposits(user string) ([]*DepositInfo, error)
	GetUserWithdrawals(user string) ([]*WithdrawalInfo, error)
	GetUserPoolBalance(poolID, user string) (*UserBalance, error)
	GetUserOwnedPools(user string) ([]*PoolInfo, error)

	// Pool deposits/withdrawals
	GetPoolDeposits(poolID string, offset, limit int) ([]*DepositInfo, int, error)
	GetPoolWithdrawals(poolID string, offset, limit int) ([]*WithdrawalInfo, int, error)
	GetPendingWithdrawals(poolID string) ([]*WithdrawalInfo, error)

	// Estimates
	EstimateDeposit(poolID string, amount math.LegacyDec) (*DepositEstimate, error)
	EstimateWithdrawal(poolID string, shares math.LegacyDec) (*WithdrawalEstimate, error)

	// Transactions
	Deposit(poolID, user string, amount math.LegacyDec) (*DepositResult, error)
	RequestWithdrawal(poolID, user string, shares math.LegacyDec) (*WithdrawalResult, error)
	ClaimWithdrawal(withdrawalID, user string) (*ClaimResult, error)
	CancelWithdrawal(withdrawalID, user string) error

	// Revenue
	GetPoolRevenue(poolID string) (*RevenueStats, error)
	GetRevenueRecords(poolID string, limit int) ([]*RevenueRecord, error)
	GetRevenueBreakdown(poolID string) (*RevenueBreakdown, error)

	// Community Pool
	CreateCommunityPool(owner string, params *CommunityPoolParams) (*PoolInfo, error)
	UpdateCommunityPool(poolID, owner string, params *CommunityPoolParams) (*PoolInfo, error)
	GetPoolHolders(poolID string) ([]*HolderInfo, error)
	GetPoolPositions(poolID string) ([]*PositionInfo, error)
	GetPoolTrades(poolID string, limit int) ([]*PoolTradeInfo, error)
	GetInviteCodes(poolID, owner string) ([]*InviteCode, error)
	GenerateInviteCode(poolID, owner string) (*InviteCode, error)
	PlacePoolOrder(poolID, owner, marketID, side string, size, price, leverage math.LegacyDec) (*PoolOrderResult, error)
	ClosePoolPosition(poolID, owner, positionID string) (*PoolCloseResult, error)
	PausePool(poolID, owner string) error
	ResumePool(poolID, owner string) error
	ClosePool(poolID, owner string) error
}

// Data types for RiverPool service

type PoolInfo struct {
	PoolID              string `json:"pool_id"`
	PoolType            string `json:"pool_type"` // "foundation", "main", "community"
	Name                string `json:"name"`
	Description         string `json:"description"`
	Status              string `json:"status"` // "active", "paused", "closed"
	TotalDeposits       string `json:"total_deposits"`
	TotalShares         string `json:"total_shares"`
	NAV                 string `json:"nav"`
	HighWaterMark       string `json:"high_water_mark"`
	CurrentDrawdown     string `json:"current_drawdown"`
	DDGuardLevel        string `json:"dd_guard_level"` // "normal", "warning", "critical"
	MinDeposit          string `json:"min_deposit"`
	MaxDeposit          string `json:"max_deposit"`
	LockPeriodDays      int64  `json:"lock_period_days"`
	RedemptionDelayDays int64  `json:"redemption_delay_days"`
	DailyRedemptionLimit string `json:"daily_redemption_limit"`
	SeatsAvailable      int64  `json:"seats_available,omitempty"`
	Owner               string `json:"owner,omitempty"` // Community pool only
	CreatedAt           int64  `json:"created_at"`
	UpdatedAt           int64  `json:"updated_at"`
}

type PoolStats struct {
	PoolID           string `json:"pool_id"`
	TotalDeposits    string `json:"total_deposits"`
	TotalWithdrawals string `json:"total_withdrawals"`
	TotalRevenue     string `json:"total_revenue"`
	NAV              string `json:"nav"`
	APY30d           string `json:"apy_30d"`
	APY7d            string `json:"apy_7d"`
	MaxDrawdown      string `json:"max_drawdown"`
	SharpeRatio      string `json:"sharpe_ratio"`
	HolderCount      int    `json:"holder_count"`
}

type NAVPoint struct {
	Timestamp int64  `json:"timestamp"`
	NAV       string `json:"nav"`
}

type DDGuardState struct {
	PoolID          string `json:"pool_id"`
	Level           string `json:"level"` // "normal", "level1", "level2", "level3"
	CurrentDrawdown string `json:"current_drawdown"`
	HighWaterMark   string `json:"high_water_mark"`
	TriggerHistory  []DDGuardTrigger `json:"trigger_history"`
}

type DDGuardTrigger struct {
	Timestamp int64  `json:"timestamp"`
	Level     string `json:"level"`
	Drawdown  string `json:"drawdown"`
	Action    string `json:"action"`
}

type DepositInfo struct {
	DepositID   string `json:"deposit_id"`
	PoolID      string `json:"pool_id"`
	User        string `json:"user"`
	Amount      string `json:"amount"`
	Shares      string `json:"shares"`
	NAVAtDeposit string `json:"nav_at_deposit"`
	Status      string `json:"status"`
	LockedUntil int64  `json:"locked_until,omitempty"`
	CreatedAt   int64  `json:"created_at"`
}

type WithdrawalInfo struct {
	WithdrawalID    string `json:"withdrawal_id"`
	PoolID          string `json:"pool_id"`
	User            string `json:"user"`
	Shares          string `json:"shares"`
	EstimatedAmount string `json:"estimated_amount"`
	ActualAmount    string `json:"actual_amount,omitempty"`
	Status          string `json:"status"` // "pending", "claimable", "claimed", "cancelled"
	RequestedAt     int64  `json:"requested_at"`
	ClaimableAt     int64  `json:"claimable_at"`
	ClaimedAt       int64  `json:"claimed_at,omitempty"`
}

type UserBalance struct {
	PoolID      string `json:"pool_id"`
	User        string `json:"user"`
	Shares      string `json:"shares"`
	Value       string `json:"value"`
	UnrealizedPnL string `json:"unrealized_pnl"`
	DepositedAmount string `json:"deposited_amount"`
}

type DepositEstimate struct {
	PoolID        string `json:"pool_id"`
	Amount        string `json:"amount"`
	EstimatedShares string `json:"estimated_shares"`
	CurrentNAV    string `json:"current_nav"`
	MinDeposit    string `json:"min_deposit"`
	PointsReward  string `json:"points_reward,omitempty"`
}

type WithdrawalEstimate struct {
	PoolID          string `json:"pool_id"`
	Shares          string `json:"shares"`
	EstimatedAmount string `json:"estimated_amount"`
	CurrentNAV      string `json:"current_nav"`
	DelayDays       int    `json:"delay_days"`
	DailyLimit      string `json:"daily_limit"`
	QueuePosition   int    `json:"queue_position,omitempty"`
}

type DepositResult struct {
	DepositID   string `json:"deposit_id"`
	PoolID      string `json:"pool_id"`
	User        string `json:"user"`
	Amount      string `json:"amount"`
	Shares      string `json:"shares"`
	NAV         string `json:"nav"`
	LockedUntil int64  `json:"locked_until,omitempty"`
}

type WithdrawalResult struct {
	WithdrawalID string `json:"withdrawal_id"`
	PoolID       string `json:"pool_id"`
	User         string `json:"user"`
	Shares       string `json:"shares"`
	ClaimableAt  int64  `json:"claimable_at"`
}

type ClaimResult struct {
	WithdrawalID string `json:"withdrawal_id"`
	Amount       string `json:"amount"`
	ClaimedAt    int64  `json:"claimed_at"`
}

type RevenueStats struct {
	PoolID            string `json:"pool_id"`
	TotalRevenue      string `json:"total_revenue"`
	SpreadRevenue     string `json:"spread_revenue"`
	FundingRevenue    string `json:"funding_revenue"`
	LiquidationRevenue string `json:"liquidation_revenue"`
	Period            string `json:"period"`
}

type RevenueRecord struct {
	Timestamp int64  `json:"timestamp"`
	Type      string `json:"type"` // "spread", "funding", "liquidation"
	Amount    string `json:"amount"`
	TradeID   string `json:"trade_id,omitempty"`
}

type RevenueBreakdown struct {
	PoolID     string `json:"pool_id"`
	Spread     string `json:"spread"`
	Funding    string `json:"funding"`
	Liquidation string `json:"liquidation"`
	Total      string `json:"total"`
}

type CommunityPoolParams struct {
	Name             string `json:"name"`
	Description      string `json:"description"`
	MinDeposit       string `json:"min_deposit"`
	MaxDeposit       string `json:"max_deposit"`
	ManagementFee    string `json:"management_fee"`    // e.g., "0.02" for 2%
	PerformanceFee   string `json:"performance_fee"`   // e.g., "0.20" for 20%
	LockPeriodDays   int    `json:"lock_period_days"`
	RedemptionDelay  int    `json:"redemption_delay_days"`
	OwnerMinStake    string `json:"owner_min_stake"`   // e.g., "0.05" for 5%
	IsPrivate        bool   `json:"is_private"`
}

type HolderInfo struct {
	User           string `json:"user"`
	Shares         string `json:"shares"`
	SharePercent   string `json:"share_percent"`
	Value          string `json:"value"`
	DepositedAt    int64  `json:"deposited_at"`
}

type PositionInfo struct {
	PositionID string `json:"position_id"`
	MarketID   string `json:"market_id"`
	Side       string `json:"side"`
	Size       string `json:"size"`
	EntryPrice string `json:"entry_price"`
	MarkPrice  string `json:"mark_price"`
	PnL        string `json:"pnl"`
	Leverage   string `json:"leverage"`
}

type PoolTradeInfo struct {
	TradeID    string `json:"trade_id"`
	MarketID   string `json:"market_id"`
	Side       string `json:"side"`
	Size       string `json:"size"`
	Price      string `json:"price"`
	Fee        string `json:"fee"`
	PnL        string `json:"pnl"`
	ExecutedAt int64  `json:"executed_at"`
}

type InviteCode struct {
	Code       string `json:"code"`
	PoolID     string `json:"pool_id"`
	MaxUses    int    `json:"max_uses"`
	UsedCount  int    `json:"used_count"`
	ExpiresAt  int64  `json:"expires_at"`
	CreatedAt  int64  `json:"created_at"`
}

type PoolOrderResult struct {
	OrderID   string `json:"order_id"`
	PoolID    string `json:"pool_id"`
	MarketID  string `json:"market_id"`
	Side      string `json:"side"`
	Size      string `json:"size"`
	Price     string `json:"price"`
	Status    string `json:"status"`
	CreatedAt int64  `json:"created_at"`
}

type PoolCloseResult struct {
	PositionID string `json:"position_id"`
	PoolID     string `json:"pool_id"`
	RealizedPnL string `json:"realized_pnl"`
	ClosedAt   int64  `json:"closed_at"`
}
